### PR TITLE
MAINT: sparse: clean up `_sputils.getdtype` docstring

### LINKED ISSUE #20879: MAINT: clean up `sparse._sputils.getdtype`
The docstring of `getdtype` is very hard to understand. Provide proper parameter descriptions, and ensure the functionality matches the docstring (it does not currently, but might after #20754).

### PR DESCRIPTION
**Pull request by Searchingdays**
Some changes to the docstring of getdtype function and adding a Value error if dtype not among the supported types list.
#### What does this implement/fix?

Closes #20879

### CODE REVIEW

**Comment by perimosocordiae** [comment]
Looks like you have some unrelated changes in this branch. I'd recommend waiting until gh-20754 is merged (which should happen soon, hopefully today) and then rebase on the upstream main branch.

**Comment by lucascolley** [comment]
gh-20754 was merged if you would like to rebase!

**Review by ev-br — DISMISSED** [review]
This is still wrong, and even if were not, it should not be mixed with sparse changes.

**Comment by dschult** [comment]
I have removed the accidentally included commits and rebased after gh-20754
I also added a commit that tweaks the message and lists the supported dtypes to the doc_string.
The error message for an unsupported dtype includes this list (built dynamically), so if you'd prefer
not to have it in the doc_string I can remove it instead. 

The linter complained about an adjacent function using `==` for types so I switched to `isinstance`.
This should be ready for review.

File: scipy/sparse/_sputils.py
Comment (stefanv) [inline scipy/sparse/_sputils.py:112]: The first sentence of a docstring usually describes what the function does. What should it say in this case? I was going to make a suggestion, but I'm not sure!

File: scipy/sparse/_sputils.py
Comment (stefanv) [inline scipy/sparse/_sputils.py:114]: But the default value of `default` is None... :thinking:

**Review by stefanv — COMMENTED** [review]
Thanks for taking a look at this @lucascolley!

File: scipy/sparse/_sputils.py
Comment (dschult) [reply scipy/sparse/_sputils.py:112]: How about:
Determines and checks a dtype based on the arguments provided.

File: scipy/sparse/_sputils.py
Comment (stefanv) [reply scipy/sparse/_sputils.py:114]:
```
`If `default` is None and needed, a TypeError is raised`.
```

But, yes, something like that is good!

File: scipy/sparse/_sputils.py
Comment (stefanv) [reply scipy/sparse/_sputils.py:112]: I'm trying to document "What would I use this function for", for the benefit of a developer. E.g., "Determine common dtype between input arguments" (not quite) ...

Comment (dschult) [reply scipy/sparse/_sputils.py:112]: How about: 

```
    """Form a supported numpy dtype based on input arguments.

    Returns a valid ``numpy.dtype`` from `dtype` if not None,
    or else ``a.dtype`` if possible, or else the given `default`
    if not None, or else raise a ``TypeError``.

    The resulting ``dtype`` must be in ``supported_dtypes``:
        bool_, int8, uint8, int16, uint16, int32, uint32,
        int64, uint64, longlong, ulonglong, float32, float64,
        longdouble, complex64, complex128, clongdouble
```

**Review by dschult — APPROVED** [review]
Suggestions from @stefanv and @dschult have been included.
I approve this PR.

**Review by lucascolley — APPROVED** [review]
thanks all!

**Comment by stefanv** [comment]
Thank you!

**Pull request merged**
